merlin/models/torch/models/base.py

```diff
@@ -0,0 +1,226 @@
+#
+# Copyright (c) 2023, NVIDIA CORPORATION.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+#
+from typing import Dict, List, Optional, Sequence, Union
+
+import torch
+from pytorch_lightning import LightningModule
+from torch import nn
+
+from merlin.dataloader.torch import Loader
+from merlin.io import Dataset
+from merlin.models.torch.batch import Batch, sample_batch
+from merlin.models.torch.block import Block
+from merlin.models.torch.container import BlockContainer
+from merlin.models.torch.outputs.base import ModelOutput
+from merlin.models.torch.utils import module_utils
+from merlin.models.utils.registry import camelcase_to_snakecase
+from merlin.schema import Schema
+
+
+class Model(Block, LightningModule):
+    """
+    Merlin Model class.
+
+    The Model class extends from both the Block and LightningModule classes. It
+    allows for easy construction of models using pre-defined blocks.
+
+    Parameters
+    ----------
+    *blocks: nn.Module
+        One or more blocks that make up the core functionality of the model.
+    schema: Schema, optional
+        A Merlin schema. Default is None.
+    optimizer: torch.optim.Optimizer, optional
+        A PyTorch optimizer from the PyTorch library (or any custom optimizer
+        that follows the same API). Default is Adam optimizer.
+
+    Example usage
+    -------------
+    >>> model = Model(
+    ...    TabularInputBlock(schema),
+    ...    MLPBlock([32, 16]),
+    ...    BinaryOutput(schema.select_by_tag(Tags.TARGET).first),
+    ... )
+    ... trainer = Trainer(max_epochs=1)
+    ... with Loader(dataset, batch_size=16) as loader:
```

Comment:
Is it required to use it as a context-manager?

Reply:
No, it's probably not necessary in the torch case, but I want to promote this idiom using the context manager everywhere, because the tensorflow equivalent has a memory leak in some cases without a context manager.